File: 0436-find-right-interval/0436-find-right-interval.cpp
class Solution {
public:
    vector<int> findRightInterval(vector<vector<int>>& intervals) {
        vector<pair<int, int>> starts;
        for (int i = 0; i < intervals.size(); i++) {
            starts.push_back({intervals[i][0], i});
        }
        sort(starts.begin(), starts.end());

        vector<int> res;
        for (auto interval : intervals) {
            int index = search(interval, starts);
            res.push_back(index);
        }
        return res;
    }

    int search(vector<int>& interval, vector<pair<int, int>>& starts) {
        int left = 0, right = starts.size();
        while (left < right) {
            int mid = (left + right) / 2;
            if (starts[mid].first < interval[1]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left >= starts.size() ? -1 : starts[left].second;
    }
};

File: 0436-find-right-interval/0436-find-right-interval.py
class Solution:
    def findRightInterval(self, intervals: List[List[int]]) -> List[int]:
        n = len(intervals)
        arr = []  # stores first item and index of each interval

        for i in range(n):
            arr.append([intervals[i][0], i])
        arr.sort()

        # implement lower_bound
        res = []
        for i in range(n):
            # r should be n not n - 1 to ensure that l reach the last index.
            l, r = 0, n
            while l < r: # because r = mid not mid - 1
                mid = (l + r) // 2
                if arr[mid][0] >= intervals[i][1]:
                    r = mid
                else:
                    l = mid + 1
            x = -1
            if l < n:
                x = arr[l][1]

            res.append(x)

        return res
